Add Capsule class with inertia calculation method

scpeters opened the pull request:
This package has classes for several shapes (`Box`, `Cylinder`, `Sphere`) as well as methods in `MassMatrix3` for computing inertial properties of these shapes assuming uniform density. This pull request adds a new `Capsule` shape class that is like a `Cylinder` with hemispheres capping each end. Methods for computing the inertial properties of a `Capsule` are added to `MassMatrix3` and validated in unit tests by comparing to the [formulae used by Open Dynamics Engine](https://bitbucket.org/odedevs/ode/src/0.16.2/ode/src/mass.cpp#lines-148:153).

Update: I've removed the `MassMatrix3` methods and removed the `Quaternion` parameter from the `Capsule` class.

EricCousineau-TRI commented on include/ignition/math/MassMatrix3.hh:
```diff
@@ -1106,6 +1106,107 @@ namespace ignition
+        //    83/320 mass * radius^2 about other axis
+        //    efunda.com/math/solids/solids_display.cfm?SolidName=Hemisphere
+        // add inertia of hemispheres about their centroid
+        L(0, 0) += sphereMass * 83. / 320. * radius2;
```
While these magic numbers are traceable, they seem a bit too magical, when really it's just the parallel-axis theorem with some solids (cylinder + hemisphere)?

Rather than bake this all together (which yeah, sure, it's great for efficiency, but how often will this be called?), why not leverage ign math to do parallel-axis with the constituent inertias?
Then you don't need comments, b/c the code is clear.

scpeters replied:
I cleaned this up a bit in 0ac1488

EricCousineau-TRI commented on include/ignition/math/MassMatrix3.hh:
```diff
@@ -1106,6 +1106,107 @@ namespace ignition
+      {
+        // Check that density, _radius and _length are strictly positive
+        // and that quaternion is valid
+        if (_mat.Density() <= 0 || _length <= 0 || _radius <= 0 ||
```
(Also, the fact that you have to make an overload named `SetFromCapsuleZ` means that now you have to redefine the frames that are already defined in the Capsule class itself?)

scpeters replied:
I removed these duplicate functions in d9c8d73

EricCousineau-TRI reviewed (approved):
Looks good at a high-level. The API here seems to be highly redundant and coupled (e.g. shapes + offsets, inertia computations, etc.), but understandably it'd be hard to refactor any time soon.

Made some picky comments, but nothing blocking.

Is there also a high-level issue that this is aiming towards?
e.g. https://github.com/osrf/sdformat/issues/176

EricCousineau-TRI commented:
Also, the shapes defined in `ign-math` seem like they have a shared contract, but there doesn't seem to be any form of contract enforcement (e.g. dynamic inheritance, static inheritance / mixins with CRTP, etc.).

Am I missing something there?

scpeters commented:
> Is there also a high-level issue that this is aiming towards?
> e.g. [osrf/sdformat#176](https://github.com/osrf/sdformat/issues/176) [osrf/sdformat#376](https://github.com/osrf/sdformat/issues/376)

it's in support of https://github.com/osrf/sdformat/issues/376, along with https://github.com/osrf/sdformat/pull/389

scpeters commented:
> The API here seems to be highly redundant and coupled (e.g. shapes + offsets, inertia computations, etc.), but understandably it'd be hard to refactor any time soon.

It took me a little while to remember, but I remembered a bit of the original intent behind this design. We added the `MassMatrix3` class (started in 2.3.0 and a bit more in 2.4.0) before adding the shape classes like `Sphere` and `Cylinder` (5.0.0). The `MassMatrix3` class is intended to represent a scalar mass and a 3x3 moment of inertia matrix. If you diagonalize the inertia matrix to get the principal moments of inertia and the orientation of its principal axes, then you can use these values along with its mass to compute the dimensions and orientation of a box of uniform density that has equivalent inertial properties (see [MassMatrix3::EquivalentBox](https://github.com/ignitionrobotics/ign-math/blob/ignition-math2_2.4.0/include/ignition/math/MassMatrix3.hh#L683-L727) from 2.4.0 for reference).

That `EquivalentBox` is used to visualize the moment of inertia in Gazebo. After adding `EquivalentBox`, we added its inverse `SetFromBox` as well as `SetFromSphere` and `SetFromCylinderZ` for good measure (we also use these in gazebo). The quaternion parameter was added to `SetFromBox` for symmetry with `EquivalentBox` and to `SetFromCylinderZ` to accommodate cylinders with different axis alignments.

When the shape functions were added in 5.0.0, they included a Quaternion parameter in the constructor, and in retrospect, I think they might be confusing things more than helping. I'll remove them from the `Capsule` constructors; they'll still be in the other shape classes, but we'll need a major version bump to get rid of those

I'll think further on whether we need `MassMatrix3::SetFromCapsuleZ` functions or not; if not I'll just put the math in the `Capsule` class

scpeters commented on include/ignition/math/Capsule.hh:
```diff
@@ -0,0 +1,189 @@
+    /// length, rotational offset, and material properties. The shape is
+    /// equivalent to a cylinder with capped with hemispheres. Radius and
+    /// length are in meters. See Material for more on material properties.
+    /// By default, a capsule's length is aligned with the Z axis. The
```
I've removed the `Quaternion` from the `Capsule` class in 4de985ebddf5c104b6fce745a175d3c7a397d01e

do you think we still need to define frames in the documentation?

EricCousineau-TRI replied:
Nah, if there's just one frame, it's a lot easier to reason about locally. Thanks!

scpeters commented:
> I'll think further on whether we need `MassMatrix3::SetFromCapsuleZ` functions or not; if not I'll just put the math in the `Capsule` class

I thought about it, and we can use the `Inertial` class directly from `Capsule`, which we can't do from `MassMatrix3`, so I've done it in 0ac1488, and it's a bit more elegant

I've removed the `MassMatrix3::SetFromCapsuleZ` helper functions in d9c8d73

chapulina commented on include/ignition/math/detail/Capsule.hh:
```diff
@@ -0,0 +1,157 @@
+T Capsule<T>::DensityFromMass(const T _mass) const
+{
+  if (this->radius <= 0 || this->length <=0 || _mass <= 0)
+    return -1.0;
```
I know this is following the pattern from other shapes, but I think we should consider returning `NaN` in such cases instead of `-1`.

scpeters replied:
what do you think @azeey?

azeey replied:
I'm not confident if returning a `NaN` is good practice, but the math functions in the standard c++ library either throw an exception or return a `NaN` when the input is invalid, so I think it's reasonable to return a `NaN`.

scpeters replied:
we don't use `NaN`s much in our APIs, and I'm careful about introducing them because they can quickly spread and it can be hard to figure out where they came from. In this case, I think it's fine because this function isn't widely used

scpeters replied:
2628fabf38785957845ac985d29ea0b3cce07613

chapulina commented on include/ignition/math/detail/Capsule.hh:
```diff
@@ -0,0 +1,157 @@
+  // since centroid of hemisphere is 3/8 radius from its flat base
+  const T dz = this->length / 2. + this->radius * 3. / 8.;
+  Inertial<T> upperCap(hemisphere, Pose3<T>(0, 0, dz, 0, 0, 0));
+  Inertial<T> lowerCap(hemisphere, Pose3<T>(0, 0, -dz, 0, 0, 0));
```
I'm missing a 180 degree rotation here, but I guess it wouldn't make a difference since this is about the centroid?

scpeters replied:
yeah, it doesn't make a difference because it's about the centroid

chapulina reviewed (approved):
Looks good to me, I just have a question about the hemispheres, but I'm trusting the test that the math checks out.

scpeters commented on include/ignition/math/Capsule.hh:
```diff
@@ -0,0 +1,189 @@
+    ///
+    /// The capsule class supports defining a capsule with a radius,
+    /// length, rotational offset, and material properties. The shape is
+    /// equivalent to a cylinder with capped with hemispheres. Radius and
```
this was fixed in 4de985ebddf5c104b6fce745a175d3c7a397d01e

scpeters commented on include/ignition/math/Capsule.hh:
```diff
@@ -0,0 +1,189 @@
+    class Capsule
+    {
+      /// \brief Default constructor. The default radius and length are both
+      /// zero. The default rotational offset is
```
I removed the rotational offset in 4de985e

scpeters commented:
> Also, the shapes defined in `ign-math` seem like they have a shared contract, but there doesn't seem to be any form of contract enforcement (e.g. dynamic inheritance, static inheritance / mixins with CRTP, etc.).
> 
> Am I missing something there?

no I don't think you're missing anything. I think the shape objects have similar API's but it's not enforced

EricCousineau-TRI commented:
Aye, sounds good. Minor concern is that the `Capsule` API may now deviates from others (given that it computes its own inertia, rather than kinda doing some psuedo static-double-dispatch things).

azeey commented on include/ignition/math/Capsule.hh:
```diff
@@ -0,0 +1,159 @@
+    {
+      /// \brief Default constructor. The default radius and length are both
+      /// zero.
+      public: Capsule() = default;
```
What do you think about not providing a default constructor? That way the only time we would have invalid Capsule objects would be if the input length and radius values are invalid.

scpeters replied:
I like having the default constructor. If you want to create a shape without remembering the order of the arguments in the constructor:

~~~
Capsuled capsule;
capsule.SetLength(length);
capsule.SetRadius(radius);
~~~

if other people don't like it, we can remove it, but I'm inclined to keep it

mxgrey replied:
If I put on my pedantry hat, the "right" way to do what @scpeters is suggesting would be to have something like a `Capsule::Builder` class that lets you fill in the information necessary to build a `Capsule` object, and then you could call `build()` on it (or something similar) and it constructs a capsule for you, perhaps throwing an exception if the parameters are invalid (e.g. radius <= 0, length <= 0).

Some would (fairly) argue that's overkill for this use case, but the more I've had to deal with writing production-quality code, the more I've been feeling it's worth the effort to make every assumption explicit inside the code itself.

azeey commented on src/Capsule_TEST.cc:
```diff
@@ -0,0 +1,127 @@
+
+  // expected values based on formula used in Open Dynamics Engine
+  // https://bitbucket.org/odedevs/ode/src/0.16.2/ode/src/mass.cpp#lines-148:153
+  double ixxIyy = (1/12.0) * cylinderMass * (3*r*r + l*l)
```
I checked this against DART (https://github.com/dartsim/dart/blob/master/dart/dynamics/CapsuleShape.cpp#L123) and there is a discrepancy, but I think what you have here is correct. The referenced page (https://www.gamedev.net/tutorials/_/technical/math-and-physics/capsule-inertia-tensor-r3856/) aligns with the formula here except the `0.25*l*l` term here vs `0.5*l*l` term there. Scrolling down further, I see a comment that it should actually be `0.25*l*l`.

scpeters replied:
it shows `H^2 m / 4` in equation 12 on that page; I think there was just a transcription error in equation 14

scpeters replied:
I picked ODE because I think it's right, though it's not authoritative and doesn't have any justification. I could add a link to that gamedev.net page as well if you think that's a good reference, despite its inconsistency

azeey replied:
The ODE reference is fine. I just checked DART to verify and saw it was different.

scpeters commented on include/ignition/math/Capsule.hh:
```diff
@@ -0,0 +1,159 @@
+
+      /// \brief Set the length in meters.
+      /// \param[in] _length The length of the capsule in meters.
+      public: void SetLength(const Precision _length) const;
```
yes, I will remove this

scpeters replied:
f0419f19dde94831cbe2aed4c4599e3cbb68e65b

scpeters commented on include/ignition/math/Capsule.hh:
```diff
@@ -0,0 +1,159 @@
+      public: Capsule() = default;
+
+      /// \brief Construct a capsule with a length, radius, and optionally
+      /// a rotational offset.
```
nope! I'll remove it

scpeters replied:
f0419f19dde94831cbe2aed4c4599e3cbb68e65b

scpeters commented on include/ignition/math/Capsule.hh:
```diff
@@ -0,0 +1,159 @@
+    /// length, and material properties. The shape is equivalent to a cylinder
+    /// aligned with the Z-axis and capped with hemispheres. Radius and
+    /// length are in meters. See Material for more on material properties.
+    template<typename Precision>
```
added in f0419f19dde94831cbe2aed4c4599e3cbb68e65b

scpeters commented on include/ignition/math/Capsule.hh:
```diff
@@ -0,0 +1,159 @@
+                  const Material &_mat);
+
+      /// \brief Destructor
+      public: ~Capsule() = default;
```
sure, f0419f19dde94831cbe2aed4c4599e3cbb68e65b

scpeters commented on include/ignition/math/detail/Capsule.hh:
```diff
@@ -0,0 +1,157 @@
+{
+
+//////////////////////////////////////////////////
+template<typename T>
```
the other shape classes are using `T` in their `detail` headers. I'll leave it as `T` for now since it's so much easier to read; we can revisit if there's a more concrete example of the problem

azeey replied:
Sounds good.

scpeters commented on include/ignition/math/Capsule.hh:
```diff
@@ -0,0 +1,159 @@
+      /// \return False if computation of the mass matrix failed, which
+      /// could be due to an invalid radius (<=0), length (<=0), or density
+      /// (<=0).
+      public: bool MassMatrix(MassMatrix3<Precision> &_massMat) const;
```
that's a good idea; it makes clear that invalid `MassMatrix3` values won't be returned. I've made the change in d379a5b366422e3b86fdac865e6d8a335031a11c
